File: python/SetPEVersion.py
# -*- coding: utf-8 -*-

# 2016-10-20T16:00+08:00


import fnmatch
import glob
import itertools
import os
import re
import subprocess
import sys

import fileutil


# Match version numbers of these formats:
# 1.2.3
# 1.2.3.4
version_number_re = r'([0-9]+(?:.[0-9]+){2,3})'

# Match version numbers of this format:
incomplete_version_number_re = r'^[0-9]+(?:.[0-9]+){2}$'

# Match a line, in Changelog.txt, which contains a valid version number
version_line_re = r'^### v{0}.*$'.format(version_number_re)


def get_topmost_version_line(changelog_file):
    with open(changelog_file, 'r', encoding=fileutil.get_file_encoding(changelog_file, 'utf-8')) as fp:
        return next(filter(lambda line: re.match(version_line_re, line), fp.readlines()))


def get_version_number(version_line):
    match_res = re.match(version_line_re, version_line)
    assert match_res, 'Invalid version line'
    if match_res:
        return match_res.groups()[0]


# 1.2.3 -> 1.2.3.0
def canonicalize_version_number(version_number):
    assert re.match(r'^{0}$'.format(version_number_re), version_number), 'Invalid version number format(neither x.x.x nor x.x.x.x)'
    if re.match(incomplete_version_number_re, version_number):
        version_number += '.0'
    return version_number


def perror(*args, **kwargs):
    sys.stderr.write(*args, **kwargs)
    sys.exit(1)


def quote_path(path):
    if path.startswith('"') and path.endswith('"'):
        return path
    return '"{0}"'.format(path)


def is_dll_or_exe(file):
    assert os.path.isfile(file)
    return fnmatch.fnmatch(file, '*.dll') or fnmatch.fnmatch(file, '*.exe')

    
def _get_full_path(candidate_path, file_name):
    if candidate_path is None:
        candidate_path = ''

    if os.path.isfile(candidate_path):
        return candidate_path
    elif os.path.isdir(candidate_path):
        return os.path.join(candidate_path, file_name)
    else:
        return os.path.join(os.path.dirname(sys.argv[0]), file_name)


def _iterate_module_files_legacy(module_path):
    assert os.path.isdir(module_path)
    yield from filter(is_dll_or_exe,
                      map(lambda item: os.path.join(module_path, item),
                          os.listdir(module_path)))


def _iterate_module_files_new(module_path):
    assert os.path.isdir(module_path)
    yield from filter(is_dll_or_exe,
                      filter(os.path.isfile,
                             map(lambda item_name: os.path.join(module_path, item_name),
                                 map(lambda item: item.name,
                                     os.scandir(module_path)))))


# `os.scandir` is new in Python 3.5, and Python 3.5 needs Windows Vista or higher. 
if sys.version_info >= (3, 5):
    iterate_module_files_v1 = _iterate_module_files_new
    run_subprocess = subprocess.run
else:
    iterate_module_files_v1 = _iterate_module_files_legacy
    run_subprocess = subprocess.call


_module_patterns = '*.dll', '*.exe'


def iterate_module_files_v2(module_path):
    assert os.path.isdir(module_path)

    for pattern in _module_patterns:
        pattern = os.path.join(module_path, pattern)
        yield from glob.iglob(pattern)


def iterate_module_files_v3(module_path):
    assert os.path.isdir(module_path)

    yield from itertools.chain.from_iterable(
        glob.iglob(pattern) for pattern in map(lambda pattern: os.path.join(module_path, pattern), _module_patterns))


def main():
    """
    Usage:
      SetPEVersion.py (--module-path=<PATH>) [--changelog=FILE] [--stampver=FILE] [--debug]
      SetPEVersion.py -h | --help
      SetPEVersion.py -v | --version

    Options:
      -c FILE --changelog=FILE    Specify the full path of "Changelog.txt"
      -s FILE --stampver=FILE     Specify the full path of "StampVer.exe"
      -m PATH --module-path=PATH  Specify a single module file(DLL or EXE) or a directory that contains module files
      -d --debug                  Show more messages for debug purpose
      -h --help                   Show this help message
      -v --version                Show version message
    """

    import docopt
    import pprint


    args = docopt.docopt(main.__doc__, version='SetPEVersion v0.1.0')
    changelog = _get_full_path(args['--changelog'], 'Changelog.txt')
    stampver = _get_full_path(args['--stampver'], 'StampVer.exe')

    if not os.path.isfile(changelog):
        perror('Changelog file not found at "{0}".'.format(changelog))

    if not os.path.isfile(stampver):
        perror('StampVer.exe not found at "{0}".'.format(changelog))

    modules = []
    if args['--module-path']:
        if os.path.isfile(args['--module-path']):
            modules.append(args['--module-path'])
        elif os.path.isdir(args['--module-path']):
            modules.extend(iterate_module_files_v3(args['--module-path']))
        else:
            perror('Invalid module path "{0}": Neither an existing file nor an existing directory.'.format(args['--module-path']))
    else:
        perror('"--module-path" option is required.')

    # Get the topmost line which contains a valid version number from Changelog.txt
    topmost_version_line = get_topmost_version_line(changelog)
    version_number = canonicalize_version_number(get_version_number(topmost_version_line))

    if args['--debug']:
        print('-' * 79)
        print(args)
        print(changelog)
        print(stampver)
        print(version_number)
        pprint.pprint(modules)
        print('-' * 79)

    for module in modules:
        # Code below does work for `StampVer.exe`.
        #cmd_args = (stampver, '-k', '-f"{0}"'.format(version_number), '-p"{0}"'.format(version_number), module)
        #subprocess.run(cmd_args)
        #
        # so I have to quote those arguments all by myself
        cmd_args = ' '.join((quote_path(stampver), '-k', '-f"{0}"'.format(version_number), '-p"{0}"'.format(version_number), quote_path(module)))
        run_subprocess(cmd_args)


if __name__ == '__main__':
    main()


# References:
# Ongoing-Study/cpp/msvc_cmdline_args/msvc_cmdline_args.cpp
# [Python glob multiple filetypes](http://stackoverflow.com/questions/4568580/python-glob-multiple-filetypes)
# https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/build/set_version.py

File: ruby/SetPEVersion.rb
# encoding: GBK
# frozen_string_literal: true
#
# = SetPEVersion.rb
#
# A Ruby clone of `Ongoing-Study/python/SetPEVersion.py`.
# `SetPEVersion.py` relies on `StampVer.exe`. This implementation
# can also use `rcedit.exe` as a replacement.
#
# 2016-12-20T18:30:07+08:00


class VersionNumberSource
    # Matchs version numbers of these format:
    # 1.2.3
    # 1.2.3.4
    VERSION_NUMBER_REGEXP = /([0-9]+(?:.[0-9]+){2,3})/

    def version_string
        raise NotImplementedError, ":version_string is not implemented by #{self}"
    end

    def self.canonize_version_string(version_string)
        if !is_version_string?(version_string)
            raise ArgumentError, 'Invalid version string'
        end

        return version_string.count('.') == 3 ? version_string : version_string + '.0'
    end

    def self.is_version_string?(version_string)
        if !version_string.is_a?(String)
            raise ArgumentError, 'A version string should be a "String" object'
        end

        return VERSION_NUMBER_REGEXP =~ version_string
    end

    def self.is_canonized_version_string?(version_string)
        return is_version_string?(version_string) && version_string.count('.') == 3
    end
end


class ChangelogSource < VersionNumberSource
    def initialize(changelog_file, mode="r:UTF-8")
        @changelog_file = changelog_file
        @mode = mode
    end

    def version_string
        File.open(@changelog_file, @mode).each { |line|
            if @@version_number_line_regexp =~ line
                return $1
            end
        }

        return nil
    end

    @@version_number_line_regexp = Regexp.new("^### v#{VERSION_NUMBER_REGEXP}.*$")
end


class VersionTool
    def initialize(tool_search_path=nil)
        @tool_search_path = tool_search_path
        @tool_path = nil
    end

    def build_cli(target_pe_file, canonized_version_string)
        @tool_path = find_tool
        if @tool_path == nil
            raise RuntimeError, 'Neither "rcedit.exe" nor "StampVer.exe" was found'
        end

        if !File.exist?(target_pe_file)
            raise ArgumentError, "\"#{target_pe_file}\" doesn't exist"
        end

        if !VersionNumberSource.is_canonized_version_string?(canonized_version_string)
            raise ArgumentError, "\"#{canonized_version_string}\" is not a canonized version string"
        end

        case @tool_path
        when /rcedit.exe$/i
            return [
                @tool_path,
                target_pe_file,
                '--set-file-version',
                canonized_version_string,
                '--set-product-version',
                canonized_version_string
            ]
        when /StampVer.exe$/i
            cmd = "\"#{@tool_path}\" -k -f\"#{canonized_version_string}\" -p\"#{canonized_version_string}\" \"#{target_pe_file}\""
            return [cmd]
        end
    end

    def find_tool
        search_paths = []

        if @tool_search_path != nil && File.exist?(@tool_search_path)
            if File.directory?(@tool_search_path)
                search_paths << @tool_search_path
            else
                return @tool_search_path
            end
        end

        search_paths << '.'
        search_paths << File.dirname(Process.argv0)
        search_paths += ENV['path'].split(/;/)

        search_paths.each { |path|
            [
                'rcedit.exe',   # https://github.com/electron/rcedit
                'StampVer.exe', # http://blog.dixo.net/downloads/stampver/
            ].each { |tool_fname|
                fullpath = File.join(path, tool_fname)
                if File.exist?(fullpath)
                    return fullpath
                end
            }
        }

        return nil
    end

    def run(target_pe_file, canonized_version_string)
        if !system(*build_cli(target_pe_file, canonized_version_string))
            puts("Failed to set the version of \"#{target_pe_file}\".")
        end
    end
end


# TODO: require 'optparse'
class CLI
    def parse(args)
        if !args.is_a?(Array)
            raise ArgumentError, '"args" is supposed to be an "Array"'
        end

        @args = args

        @parse_result = {
            :help => @args.empty?,
            :version => false,
        }

        done = false
        assign_value = nil
        @args.each { |argv|
            case argv
            when '-c', '--changelog'
                if !@parse_result.has_key?(:changelog)
                    @parse_result[:changelog] = nil
                    assign_value = ->(value) { @parse_result[:changelog] = value }
                else
                    raise RuntimeError, 'Duplicate "--changelog" arugment'
                end
            when '-t', '--tool-path'
                if !@parse_result.has_key?(:tool)
                    @parse_result[:tool] = nil
                    assign_value = ->(value) { @parse_result[:tool] = value }
                else
                    raise RuntimeError, 'Duplicate "--tool-path" arugment'
                end                
            when '-m', '--module-path'
                if !@parse_result.has_key?(:modules)
                    @parse_result[:modules] = []
                end

                assign_value = ->(value) { @parse_result[:modules] << value }
            when '-h', '--help'
                @parse_result[:help] = true
                done = true
            when '-v', '--version'
                @parse_result[:version] = true
                done = true
            else
                if argv[0] == '-'
                    raise RuntimeError, "Unknown option \"#{argv}\""
                end

                if assign_value
                    assign_value.call(argv)
                end
            end

            if done then break end
        }

        if !@parse_result[:help] && !@parse_result[:version]
            if !@parse_result.has_key?(:modules)
                raise RuntimeError, 'Missing "-m"/"--module-path" option'
            elsif @parse_result[:modules].empty?
                raise RuntimeError, 'At least one module should be specified'
            else
                modules = []
                @parse_result[:modules].each { |item|
                    if !File.exist?(item)
                        raise RuntimeError, "Module \"#{item}\" not found"
                    end

                    if File.directory?(item)
                        Dir.foreach(item) { |entry|
                            entry = File.join(item, entry)
                            if !File.directory?(entry) && File.fnmatch('*.{dll,exe}', entry, File::FNM_EXTGLOB | File::FNM_CASEFOLD)
                                modules << entry
                            end
                        }
                    else
                        modules << item
                    end
                }
                @parse_result[:modules] = modules
            end

            if !@parse_result.has_key?(:tool)
                @parse_result[:tool] = nil
            elsif !@parse_result[:tool]
                raise RuntimeError, 'Missing argument for "-t"/"--tool-path" option'
            end

            if !@parse_result.has_key?(:changelog)
                @parse_result[:changelog] = 'Changelog.txt'
            elsif !@parse_result[:changelog]
                raise RuntimeError, 'Missing argument for "-c"/"--changelog" option'
            else
                if File.directory?(@parse_result[:changelog])
                    @parse_result[:changelog] = File.join(@parse_result[:changelog], "Changelog.txt")
                end

                if !File.exist?(@parse_result[:changelog])
                    raise RuntimeError, "Changelog file \"#{@parse_result[:changelog]}\" not found"
                end
            end
        end
    end

    def run(args)
        parse(args)

        if @parse_result[:help]
            puts(usage)
            return
        elsif @parse_result[:version]
            puts("SetPEVersion v0.1.0")
            return
        end

        version_tool = VersionTool.new(@parse_result[:tool])
        changelog_source = ChangelogSource.new(@parse_result[:changelog])
        version_number = VersionNumberSource.canonize_version_string(changelog_source.version_string)

        @parse_result[:modules].each { |pe|
            version_tool.run(pe, version_number)
        }
    end

    def usage
        'Usage:
  SetPEVersion (--module-path=<PATH>) [--changelog=FILE] [--tool-path=FILE]
  SetPEVersion -h | --help
  SetPEVersion -v | --version

  Options:
    -c FILE --changelog=FILE    Specify the full path of "Changelog.txt"
    -t FILE --tool-path=FILE    Specify the search path of "rcedit.exe"/"StampVer.exe"
    -m PATH --module-path=PATH  Specify a single module file(DLL or EXE) or a directory that contains module files
    -h --help                   Show this help message
    -v --version                Show version message'
    end
end


CLI.new.run(ARGV)


# References:
# [What's the point of ARGV in Ruby?](http://stackoverflow.com/questions/13329132/whats-the-point-of-argv-in-ruby)
# Ongoing-Study/python/SetPEVersion.py
# https://github.com/ryangreenberg/urban_dictionary
